### DISCUSSION TITLE
Explain the Heuristics and the Naïve Bayesian Classification

### DISCUSSION THREAD

**Discussion by zzj0402** [github-linguist/linguist#6286] [General]
Hi all,
I am working on a code classification task for my doctorate research. Github Linguist offers an exciting solution to detect coding language as explained here https://github.com/github/linguist/blob/master/docs/how-linguist-works.md . However, I need more details on the Heuristics and on the Naïve Bayesian Classification. Please elaborate:

- What is the Heuristics?
- What are the feature vectors for the Naïve Bayesian Classification?
- What is the training data if any?
- What is the result of the evaluation if any?

Cheers from New Zealand.

**Comment by lildude** [comment]
> * What is the Heuristics?

They're regular expressions. See [`lib/linguist/heuristics.yml`](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.yml) which are then used by [`lib/linguist/heuristics.rb`](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.rb) to find matches in the file in question.

> * What are the feature vectors for the Naïve Bayesian Classification?
> * What is the training data if any?

You can answer both of these by looking at the code in the classifier [`lib/linguist/classifier.rb`](https://github.com/github/linguist/blob/master/lib/linguist/classifier.rb)

> * What is the result of the evaluation if any?

A programming language 😁 

In short the classification works like a funnel. We passing in all languages Linguist knows about ([`lib/linguist/languages.yml`](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml)) and then use different [strategies](https://github.com/github/linguist/blob/c1c34e5260797b4d598f5ec76f19723bfc5a1894/lib/linguist.rb#L62-L71) to whittle that list down to a single language. The only "machine learning" involved is in the classifier which is a last resort and does quite a mediocre job at it.

We have a PR open to switch out the bayesian classifier for a centroid-based classifier in https://github.com/github/linguist/pull/5103 which will come at some point in the future.
